fix: Organization.add_user INTEGRITY error

mohebmithani opened the pull request:
since this project uses 'add_user' method of 'django-organizations' package, I catched the specific integrity error so that I don't interfere with the imported package internals.

Fixes #360

mohebmithani commented:
@nemesifier This project uses the functionality of `django-organizations` package and if you look into the internals of that package you'll find that `add_user` uses a protected method. So I thought it is better to handle exception then create a user_add method again. Let me know if this is the right approach. If this is right then I can further create a test case for this fix if you want

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,29 @@ class Organization(BaseOrganization, AbstractOrganization):
+            # if ``super().add_user()`` again tries to create
+            # organization owner then catch the error and returns 
+            # the organization user created
+            if "UNIQUE constraint failed: openwisp_users_organizationowner.organization_user_id" in str(e):
```
I am not sure this would work in all databases

mohebmithani replied:
Do you want me to just except integrity error not check the exact message?

nemesifier replied:
I just thought maybe we could find a better way, see here:

https://github.com/bennylope/django-organizations/blob/2dd044956cf03dad9c8c4954ba325d2769fe3643/src/organizations/abstract.py#L96C1-L117

What if we override the `create` method of our org owner model to verify whether an owner already exists before creating a new one?

mohebmithani replied:
Overriding was my first solution but I thought may be we wanted to use django-organizations methods that's why i implemented this. Let me override this and send you PR, thanks.

nemesifier reviewed (changes requested):
Thanks @mohebmithani! 

A failing test is missing though. When we fix bug we write a test case which replicates the issue first (eg: fails without the fix). Without that we cannot be confident that 1. the bug exists, 2. has been solved, 3. it won't come back in the future due to inadverted changes to the code.

mohebmithani commented:
@nemesifier yes i justed wanted to verify the approach. I'll implement the test case.

mohebmithani commented:
@nemesifier I thought about overriding `create` method of org owner. I don't think it would be a good solution. Because `create` method would then check if owner exists or not, If exists then it won't create another object. 

Let's say if someone wanted to add some feature which depends on the object creation of `owner`? if `objects.create` is successful then do **this** otherwise do **that**. I don't think create method should be changed just for this usecase. This error is occuring in `add_user` method then I think only this method should be changed. 

I've changed the implementation of `add_user` a bit. Now it will just create `org user` leaving creation of `org owner` to the signals. I've implemented test as well and tested every test cases. Let me know if more changes are required.

mohebmithani commented:
@nemesifier just topping up this message as a reminder. Actually i've selected this issue as part of 'Hacktoberfest' and I would realy appreciate if you can look into this before the event ends. Apologies and Thanks in advance

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,24 @@ class Organization(BaseOrganization, AbstractOrganization):
     class Meta(AbstractOrganization.Meta):
         swapper.swappable_setting('openwisp_users', 'Organization')
 
+    def add_user(self, user, **kwargs):
```
please move this to `openwisp_users/base/models.py`.

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,24 @@ class Organization(BaseOrganization, AbstractOrganization):
+        """
+        overriding this method of ``AbstractOrganization``
+        to remove organization owner creation logic as it is
+        being handled separately by a signal of this project.
```
Let's shorten this docstring, something like the following will work:

We override this method from the upstream dependency to
skip the creation of the organization owner, which we perform
automatically via a signal receiver.
Without this change, the add_user method would throw IntegrityError.

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,24 @@ class Organization(BaseOrganization, AbstractOrganization):
+        being handled separately by a signal of this project.
+        """
+
+        users_count = self.users.all().count()
```
since we are at it, we can use `exists()` which is more efficient, will return `True` or `False`.

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,24 @@ class Organization(BaseOrganization, AbstractOrganization):
+            is_admin = True
+        
+        OrganizationUser = swapper.load_model('openwisp_users', 'OrganizationUser')
+        org_user = OrganizationUser.objects.create(
+            user=user, organization=self, is_admin=is_admin
+        )
```
let's just return here, eg:

```suggestion
        return OrganizationUser.objects.create(
            user=user, organization=self, is_admin=is_admin
        )
```

nemesifier commented on openwisp_users/models.py:
```diff
@@ -25,6 +26,24 @@ class Organization(BaseOrganization, AbstractOrganization):
+        org_user = OrganizationUser.objects.create(
+            user=user, organization=self, is_admin=is_admin
+        )
+
+        return org_user
+
```
```suggestion
```

nemesifier commented on openwisp_users/tests/test_models.py:
```diff
@@ -333,3 +333,15 @@ def test_organization_user_string_representation(self):
+    def test_add_user_method(self):
+        """
+        test adding user does not throw `INTEGRITY` error due to double owner creation
+        logic in `add_user` method and a signal trigger.
```
Please remove this docstring as it's a redundant with the docstring of the `add_user` method, we do not need it (we can track the reason for this change by looking at the commit history and looking at this PR).

nemesifier commented on openwisp_users/models.py:
```diff
@@ -1,5 +1,6 @@
 import swapper
 from django.contrib.auth.models import Group as AbstractGroup
+from django.db import IntegrityError
```
Please make sure to run `openwisp-qa-format` and then `./run-qa-checks` to make sure the QA checks are passing.

nemesifier reviewed (changes requested):
Thanks @mohebmithani, see my comments below.

mohebmithani commented:
@nemesifier Thanks for the update. I've incorporated all the feedbacks and have ran all the checks.

nemesifier commented on openwisp_users/base/models.py:
```diff
@@ -206,6 +206,22 @@ def __str__(self):
+        Without this change, the add_user method would throw IntegrityError.
+        """
+
+        if self.users.all().exists():
```
Are you sure of this?
Wasn't the core flagging the user as admin if there were no user for the org?
While here it's flagging the user as admin if there's already users for the org.

Shouldn't it be `if not self.users.all().exists()`?

mohebmithani replied:
@nemesifier really sorry for this silly mistake. Have add not in the if statement.

mohebmithani commented:
@nemesifier did you get a chance to look at it? hopefully now everything will be on thier place.

nemesifier commented:
Thanks @mohebmithani :+1:
